feat: validate coordinate names in one place, for every backend

ikrommyd opened the pull request:
:robot: _AI text below_ :robot:

## Description

Rebased on `main` and reimplemented. Solves the half of #660 that vector can fix on its own.

### The problem

Each backend had its own rules for which coordinate names make a vector — `obj()`, the `VectorObject*D` constructors, `array()`, the six `__array_finalize__`, `_check_names`, the `VectorSympy*D` constructors, and the Numba `vector.obj` — and they disagreed. All of these work on `main` and silently build the wrong vector:

```python
vector.array({"pt": ..., "phi": ..., "eta": ..., "mass": ..., "energy": ...})  # energy wins, mass is dead
vector.array({"x": ..., "y": ..., "rho": ..., "phi": ...})                     # rho, phi are dead
vector.array({"x": ..., "y": ..., "z": ..., "eta": ...})                       # eta is dead
vector.zip({"rho": ..., "phi": ..., "pt": ...})                                # pt is dead
vector.VectorObject2D(x=1, px=2, y=3)                                          # x == 2
```

### The fix

`_methods._check_coordinate_names` is now the only implementation. It returns `(is_momentum, dimension, coordinates, extra)` or raises, and takes the dimension and momentum-ness a constructor is restricted to, so it serves both `vector.obj` (deduces both) and `MomentumNumpy4D` (fixes both). Every backend calls it, and the error messages come from one table instead of the nine hand-written copies. Source diff: **−1016/+567**.

The awkward behaviors validate through `__awkward_validation__` (awkward#3710, needs `awkward>=2.8.11`; older versions skip it), so what never went through a vector constructor is caught too:

```python
>>> ak.zip({"pt": pt, "eta": eta, "phi": phi, "mass": mass, "energy": energy}, with_name="Momentum4D")
TypeError: MomentumArray4D with fields [...]: specify t= or tau=, but not more than one

>>> jets["rho"] = rho
TypeError: MomentumArray4D with fields [...]: 'pt' and 'rho' both map to 'rho'
```

Subclassed behaviors (coffea's) inherit it. Cost is ~0.5 µs per array creation against ~130 µs for one `v.pt` on 100k entries; `_check_coordinate_names` is `lru_cache`d on the field-name tuple.

### Intended breaks

1. A field that is a coordinate under another name is no longer carried as payload. Ordinary extra fields (`charge`, `btag`) are unaffected.
2. Generic vectors reject momentum-aliases: `VectorObject2D(px=1, py=2)` used to return a *non*-momentum vector while `vector.obj(px=1, py=2)` returns `MomentumObject2D`. Momentum classes still take generic names.
3. An array whose fields stop describing its record name is rejected — `del jets["mass"]`, or a column-projected form that kept `__record__` but dropped coordinates. On `main` those gave a half-broken vector that failed at first coordinate access.
4. Error messages changed. No test asserted on their text.

### Tests

`tests/test_pr_659.py`: all 511 subsets of the generic names and all 222 combinations with momentum-aliases, per backend, against a reference implementation written independently of the one under test; plus the record names accepted and rejected by `__awkward_validation__`, and the numba path. 3969 tests in ~5 s, 100% coverage of the new code.

Coffea's validation and this agree on all 48,597 combinations I compared, and coffea's suite passes against this branch unchanged (its own hooks shadow vector's, so nothing there needs to change).

## Checklist

- [x] Have you followed the guidelines in our Contributing document?
- [x] Have you checked to ensure there aren't any other open Pull Requests for the required change?
- [x] Does your submission pass pre-commit? (`$ pre-commit run --all-files` or `$ nox -s lint`)
- [x] Does your submission pass tests? (`$ pytest` or `$ nox -s tests`)
- [x] Does the documentation build with your changes? (`$ cd docs; make clean; make html` or `$ nox -s docs`)
- [x] Does your submission pass the doctests? (`$ pytest --doctest-plus src/vector/` or `$ nox -s doctests`)

Saransh-cpp commented on src/vector/backends/object.py:
```diff
@@ -3211,7 +3211,7 @@ def obj(**coordinates: float) -> VectorObject:
     if "E" in coordinates:
```
Should this also have the `and "t" not in generic_coordinates` condition?

ikrommyd replied:
Isn't this unnecessary? This is the first if condition where `generic_coordinates` might be populated with `t`. So `and "t" not in generic_coordinates` is a useless check no? Same goes for `tau` in your other comment below.

Saransh-cpp commented on src/vector/backends/object.py:
```diff
@@ -3220,7 +3220,7 @@ def obj(**coordinates: float) -> VectorObject:
     if "M" in coordinates:
```
Ditto, for `tau`

Saransh-cpp commented on tests/backends/test_coordinate_validation.py:
Given that these tests are related to constructing vectors, we should move them to the appropriate test backend files. The object ones can go in:

https://github.com/scikit-hep/vector/blob/e374915390e1a45f972397de053e3b788938fb81/tests/backends/test_object.py#L14

The NumPy ones (we don't have much constructor tests for NumPy at the moment, so this is perfect 😅):

https://github.com/scikit-hep/vector/blob/e374915390e1a45f972397de053e3b788938fb81/tests/backends/test_numpy.py#L85

The Awkward ones (same as NumPy):

https://github.com/scikit-hep/vector/blob/e374915390e1a45f972397de053e3b788938fb81/tests/backends/test_awkward.py#L161

ikrommyd replied:
Sure. I just thought that maybe it should be a new test because it's like 1K lines but yeah I can move them under their respective backend tests.

ikrommyd replied:
Any final suggestion here? To me, putting 1K lines under `test_issues.py` is not a great idea. My understanding is that this file is for small individual issues. The only refactoring that I think may be good here is just splitting this 1K file into multiple files per backend. Any takes?

Saransh-cpp reviewed (changes requested):
Good catch, @ikrommyd! Thanks for working on this!!

Please see my comments below:

pfackeldey commented:
> I think @pfackeldey might be able to help with this - can we add constructor checks to our custom Awkward behaviors?

I'm not sure about the following:
I think you can't really change the way `__init__` works through behaviors because awkward updates it's base class _after_ the constructor has been run.

If my understanding here is wrong, you _can_ do the following:
```python
class MyBehavior:
  def __init__(self, *args, **kwargs):
    super().__init__(*args, **kwargs)
    # check
    if set(self._layout.fields) != {"pt", "eta", "phi", "mass"}:
      raise ValueError('got different fields than this behavior expected')
```

However, @ikrommyd and me talked about this and these constructor checks are indeed helpful, but do not prevent to break the behaviors. What would instead prevent this thoroughly would be:
1. Awkward Arrays are fully immutable (you can't even set/update/delete fields)
2. we disallow getattr to retrieve fields so it's clear to the user if they invoke a property or access a field
3. vector constructors are only allowed to have access to certain fixed fields

(1.) and (3.) solves this issue entirely I think, (2.) is in addition to make people aware of when they invoke the `mass` property and when they access the `mass` field.

This is of course major break in API, so I'm not sure if we can/want to do this. I'm adding this here to write down the fix that solves this issue thoroughly. 
There could be ways to enforce this immutability wrt fields per array instance. Then we can preserve the current functionality in general awkward, but vector could enforce true immutability which does not allow to update/add/del fields to arrays with vector behaviors.
About (2.) I don't think there's really a way to solve this except for a API breakage in awkward. But this _is not_ the issue, it's just hard to understand _when_ one accesses a property and when a field through dot-access.

I could add a way in awkward-array to make them truly immutable that you can define through behaviors in vector. And in theory I could also think of a way to make post init constructor checks in awkward arrays possible for behaviors. 

If you want @ikrommyd & @Saransh-cpp I can prepare a PR and you can have a look? Or what do you think?

ikrommyd commented:
Hi @eduardo-rodrigues, it's not trivial to catch when you do `ak.zip` though and you don't use any of the vector constructor methods. Because `ak.zip` does not know what `Momentum4D` means. It does not know or care what behaviors it carries over. More than that, we can hard-code the vector behavior checks explicitly in awkward-array sure but that's not the full story. You can subclass and define your own behaviors (coffea does that a lot). So then you can do `with_name="PtEtaPhiMCandidate"` which is something that coffea defines. So what behaviors do we check? We can't hard-code all the user-defined stuff? And `with_name`, all it does is it assigns a `__record__` parameter to the array's layout to a string. It is completely blind as to what this implies.

ikrommyd commented:
Did a full write-up here of the issue: https://github.com/scikit-hep/vector/discussions/660
This PR solves part of it. For further discussion, I suggest we move there to not make the PR discussion messy.

pfackeldey commented:
Linking my proposed solution here as well: https://github.com/scikit-hep/vector/discussions/660#discussioncomment-14868529

ikrommyd commented on src/vector/backends/numpy.py:
```diff
@@ -2071,6 +2071,168 @@ def __setitem__(self, where: typing.Any, what: typing.Any) -> None:
+
+    Raises TypeError if duplicate or conflicting coordinates are detected.
+    """
+    complaint1 = "duplicate coordinates (through momentum-aliases): " + ", ".join(
```
is there a suggestion regarding where to put these complaint strings?

ikrommyd commented on src/vector/backends/numpy.py:
```diff
@@ -2138,6 +2300,9 @@ def array(*args: typing.Any, **kwargs: typing.Any) -> VectorNumpy:
     is_momentum = any(x in _repr_momentum_to_generic for x in names)
 
+    # Validate coordinates using dimension-guard pattern (same as awkward _check_names)
+    _validate_numpy_coordinates(names)
```
I've moved calling this function inside the `__array_finalize__` of the 2d, 3d, and 4d classes. The tests pass so I'll need to play with it a bit more in a terminal in case there's something I'm not remembering here.

ikrommyd replied:
Oh..now I think I remember. Look at this
https://github.com/scikit-hep/vector/blob/776f405a180f6060671072bc9106eccb60abca7e/src/vector/backends/numpy.py#L2160-L2169
It makes a decision whether it is momentum or not and which class constructor to use based on the field names. That's why I think I wanted it in the `vector.array` implementation and not in the finalize. I think it's best to validate the fields first and then decide which constructor to use. I think it's best to move it back to where it was. I have reverted the change.

ikrommyd committed:
feat: validate coordinate names in one place, for every backend

Each backend had its own rules for which coordinate names make a vector, and
they disagreed. These all worked on main and silently built the wrong vector:

    vector.array({"pt": ..., "phi": ..., "eta": ..., "mass": ..., "energy": ...})
    vector.array({"x": ..., "y": ..., "rho": ..., "phi": ...})
    vector.zip({"rho": ..., "phi": ..., "pt": ...})
    vector.VectorObject2D(x=1, px=2, y=3)

_methods._check_coordinate_names is now the only implementation; object, numpy,
awkward, sympy, and numba all call it, and the error messages come from one
table instead of nine copies.

The awkward behaviors validate through __awkward_validation__ too, so
ak.zip(..., with_name="Momentum4D") and jets["rho"] = ... are caught, which is
the part of #660 that vector can fix on its own. Needs awkward>=2.8.11;
older versions skip it.

Two intended breaks: a field that is a coordinate under another name is no
longer carried as payload, and generic vectors reject momentum-aliases.

tests/test_pr_659.py covers all 511 subsets of the generic names and all 222
alias combinations, per backend, against a separate reference implementation.

ikrommyd commented:
@henryiii send an agent over for review. I did the rebasing and consolidation with a coder and an adverserial reviewer myself. The goal was to have the validation only in one place for all backends and to have 100% test coverage.
